core/src/main/scala/org/apache/spark/deploy/master/Master.scala

```diff
@@ -126,27 +131,92 @@ private[deploy] class Master(
   }
 
   // Alternative application submission gateway that is stable across Spark versions
-  private val restServerEnabled = conf.getBoolean("spark.master.rest.enabled", true)
-  private var restServer: Option[StandaloneRestServer] = None
-  private var restServerBoundPort: Option[Int] = None
+  private val restSubmissionServerEnabled = conf.getBoolean("spark.master.rest.enabled", true)
+  private var restSubmissionServer: Option[StandaloneRestServer] = None
+  private var restSubmissionServerBoundPort: Option[Int] = None
+
+  private var rpcSubmissionEndpointRef: Option[RpcEndpointRef] = None
+
+  object RPCSubmissionEndpoint extends ThreadSafeRpcEndpoint {
```

Comment:
it's not really clear: why do you need this? why can't you just use the Master itself, since it already handles all these messages?

At the very least, this needs a comment (in the code) explaining that. Also, nested classes are generally declared at the bottom of the enclosing class.